File: IntegerToWord.java
// Time O(1) because we are finding triplet and each triplet running max 3 time. Number of triplet in 2^31 -1 is limited. 
// Space O(1)
public class IntegerToWord {

    String[] below20 ={"","One","Two","Three","Four","Five","Six","Seven","Eight","Nine", "Ten","Eleven","Twelve","Thirteen","Fourteen",
                       "Fifteen","Sixteen","Seventeen","Eighteen","Nineteen"};
  String[] tens ={"","Ten","Twenty","Thirty","Forty","Fifty","Sixty","Seventy","Eighty","Ninety"};
  
  String[] Thousands = {"","Thousand","Million","Billion"};
  
  String result;
  
  public String numberToWords(int num) {
      result ="";
      if(num==0)
          return "Zero";
     int i=0; 
      while(num > 0)    
      {
          int triplet = num %1000;
          if(triplet!=0)
               result = helper(triplet).trim()+" "+ Thousands[i]+" "+result;
          i++;
          num=num/1000;
      }
      
      return result.trim();
  }
  
  public String helper(int num){
    if(num==0)
        return "";
     else if(num<20){
         return below20[num]+" ";
     }
     else if(num < 100){
         return tens[num/10]+" "+helper(num%10);
     }
     else{
        return below20[num/100]+" Hundred "+helper(num%100);
      }    
        
     }  
  
}

File: IntegerToWord.py
#All TC passed on leetcode

class Solution:
    def numberToWords(self, num: int) -> str:

        #Here we take triplets of given number. Each triplet has 3 digits and each digit value is computed using the predefined values in tens and below_20 list. 
        #Time complexity - O(n) -> O(1) where n is no. of digits as no. of digits is max 12 for billion hence constant
        #Space complexity - O(1)

        thousands = ["", "Thousand", "Million", "Billion"]
        tens = ["", "Ten", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]
        below_20 = ["", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"]

        def helper(n):
            if n==0:
                return ""

            if n<=19:
                return below_20[n] + " "

            elif 20<=n<=99:
                return tens[n//10] + " " + helper(n%10)

            elif 100<=n<=999:
                return below_20[n//100] + " Hundred " + helper(n%100)


        res = ""
        i=0
        if num==0:
            return "Zero"

        while num>0:
            cur = num%1000
            val = helper(cur)
            if val!="":
                res = val + thousands[i] + " " + res
            i+=1
            num = num//1000
        return res.strip()
        